### PR TITLE
Release v0.0.119: JSON typed accessors (#366)

### LINKED ISSUE #366: JSON typed accessors: json_as_string, json_get_number, etc.
## Problem

Working with JSON API responses in Vera requires a two-level pattern match every time you want a typed value out of an object field. From a real HTTP+JSON program:

```vera
private fn extract_float(@Option<Json> -> @String)
  effects(pure)
{
  match @Option<Json>.0 {
    None -> "N/A",
    Some(@Json) -> match @Json.0 { JNumber(@Float64) -> float_to_string(@Float64.0), _ -> "N/A" }
  }
}
```

This pattern — unwrap the `Option`, then match the `Json` constructor — is boilerplate that every JSON API consumer will write, differing only in which constructor (`JNumber`, `JString`, `JBool`) they match on. The current stdlib (`json_get` returning `Option<Json>`) gives you the field, but leaves the last mile to the user.

## Proposed additions

**Layer 1 — type coercion accessors** (Json → Option<T>):

```vera
json_as_string(@Json.0)   -- Option<String>   — Some if JString, None otherwise
json_as_number(@Json.0)   -- Option<Float64>  — Some if JNumber, None otherwise
json_as_bool(@Json.0)     -- Option<Bool>     — Some if JBool,   None otherwise
json_as_int(@Json.0)      -- Option<Int>      — Some if JNumber (applies float_to_int), None otherwise
json_as_array(@Json.0)    -- Option<Array<Json>>        — Some if JArray, None otherwise
json_as_object(@Json.0)   -- Option<Map<String, Json>>  — Some if JObject, None otherwise
```

**Layer 2 — compound field accessors** (Json, String → Option<T>):

```vera
json_get_string(@Json.0, "key")   -- Option<String>
json_get_number(@Json.0, "key")   -- Option<Float64>
json_get_bool(@Json.0, "key")     -- Option<Bool>
json_get_int(@Json.0, "key")      -- Option<Int>
json_get_array(@Json.0, "key")    -- Option<Array<Json>>
```

These are `json_get` + the corresponding `json_as_*` composed — which is what 90% of real API-consuming code actually wants.

## Impact

The `display_weather` function from a real Cambridge weather API client shrinks from five dedicated helper functions (each wrapping the two-level match) to five single-line field extractions. Composable with existing `Option` combinators (`option_map`, `option_unwrap_or`).

## Implementation

Pure Vera prelude functions — no new WASM host imports required. Each `json_as_*` is a single match expression; each `json_get_*` chains `json_get` + `option_flat_map` + `json_as_*`.

## Naming

`json_as_*` follows the coercion-without-panic convention (Rust's `.as_str()`, Swift's optional casts). `json_get_*` mirrors the compound-accessor pattern common in config-reading libraries.

### PR DESCRIPTION
**Pull request by aallan**
## Summary

Eleven new JSON accessor functions for the built-in `Json` ADT — the ergonomic polish layer that eliminates the two-level pattern-match boilerplate every JSON API consumer would otherwise write. First non-WASM-translator stdlib addition of Stage 11: these are **pure-Vera prelude functions**, not WASM translators.

### Layer 1 — type-coercion accessors (Json → Option\<T\>)

- `json_as_string`, `json_as_number`, `json_as_bool`, `json_as_int`, `json_as_array`, `json_as_object`
- `Some` when the Json's constructor matches, `None` otherwise. Disjoint: at most one returns `Some` for any given Json.
- `json_as_int` guards `float_to_int` against its trap on NaN and infinity via `float_is_nan` / `float_is_infinite`, returning `None` for those inputs. Overflow (|f| ≥ 2^63) still traps — a deliberate choice to keep the body simple.

### Layer 2 — compound field accessors (Json, String → Option\<T\>)

- `json_get_string`, `json_get_number`, `json_get_bool`, `json_get_int`, `json_get_array`
- Each collapses `json_get(obj, key)` and the matching `json_as_*` coercion into one call. Missing field AND wrong-typed field both return `None` — exactly what 90% of real API-consuming code wants.
- No `json_get_object`: chained field access is handled by `json_get` returning `Option<Json>`, then recursing.

### Before / after

```vera
-- Before (the pattern this PR kills):
match json_parse(response) {
  Err(@String) -> Err(@String.0),
  Ok(@Json) -> match json_get(@Json.0, "x") {
    None -> Err("missing x"),
    Some(@Json) -> match @Json.0 {
      JNumber(@Float64) -> Ok(float_to_int(@Float64.0)),
      _ -> Err("x is not a number")
    }
  }
}

-- After:
match json_parse(response) {
  Err(@String) -> Err(@String.0),
  Ok(@Json) -> match json_get_int(@Json.0, "x") {
    Some(@Int) -> Ok(@Int.0),
    None -> Err("x missing or not a number")
  }
}
```

## Implementation

Pure-Vera prelude functions, not WASM emitters:

- `vera/prelude.py` `_JSON_COMBINATORS` extended with all 11 bodies as match expressions over `Json`. The compiler already injects the prelude into every module that references `Json` values — no detection changes needed beyond adding the 11 new names to the detection set and the injection-site shadow-check set.
- `vera/environment.py` registers the 11 new `FunctionInfo` entries. Layer-1 accessors come from a tight Python loop (six identical `Json → Option<T>` shapes).
- No WASM translator code added. No new host imports. Compiled modules that don't reference `Json` stay byte-identical.

One oddity worth calling out: my first pass modelled `json_as_int` as `match float_to_int(...) { Ok(@Int) -> Some(...), Err(_) -> None }` — but `float_to_int` actually returns `Int` and traps on NaN/Inf, not `Result<Int, String>`. Caught the WASM type-mismatch at smoke-test time (`expected i32, found i64`) and rewrote to use the `float_is_nan || float_is_infinite` guard. The spec entry flags this asymmetry.

## Tests

- **19 new unit tests** (`TestJsonTypedAccessors` in `tests/test_codegen.py`): every Layer-1 accessor (matching + mismatched constructors), every Layer-2 accessor (hit + missing + wrong type), edge cases (`json_as_int` NaN, infinity, toward-zero truncation for negatives), disjointness invariant.
- **2 new browser parity tests** (`TestBrowserJsonAccessors`): exhaustive Layer-1 + Layer-2 sweeps.
- **New conformance program** `tests/conformance/ch09_json_accessors.vera` (level: `run`) with 15 test functions.
- **`examples/json.vera` rewritten** to use the accessors — the weather-API scenario that previously needed a helper-per-field pattern now uses Layer-2 accessors directly.

## Docs swept

- `spec/09-standard-library.md` — new §9.6.23 with full signatures, Layer-1 / Layer-2 distinction, NaN-guard note, and verification tier classification.
- `SKILL.md` — new "Typed accessors (Layer 1 and Layer 2)" subsection after the existing Json section, plus the naming-convention paragraph extended with `json_as_` / `json_get_` prefixes.
- `CHANGELOG.md` — v0.0.119 entry with separate Added / Implementation / Example rewrite / Tests sections.
- `HISTORY.md` — Stage 11 row added; tagged-release total bumped 118 → 119.
- `ROADMAP.md` — #366 removed from short-term queue and Phase 4c; replaced with **#475 (WASM call translator bug cleanup)** as the new top priority for the next PR.
- Count bumps across TESTING / CLAUDE / AGENTS / FAQ / README / vera-README: **3,527 tests, 80 conformance programs, 164 built-in functions**.

## Release prep

- Version bumped 0.0.118 → 0.0.119 across `vera/__init__.py`, `pyproject.toml`, `docs/index.html`, `README.md`, `uv.lock`.
- Tag `v0.0.119` placed on the feature branch; will be force-moved to main HEAD after merge.

## Test plan

- [x] 19 new unit tests pass (`pytest tests/test_codegen.py::TestJsonTypedAccessors`)
- [x] 2 new browser parity tests pass
- [x] All 80 conformance programs pass
- [x] All 32 examples pass check + verify
- [x] Full pytest suite green (3,513 passed)
- [x] mypy clean
- [x] All 14 doc validation scripts green

Closes #366

## Summary by CodeRabbit

* **New Features**
  * Added 11 JSON typed accessors: six json_as_* coercions and five json_get_* field accessors. json_as_int returns None for NaN, ±infinity and out‑of‑range floats.

* **Documentation**
  * Spec, changelog, README, roadmap and guides updated; conformance/test counts bumped to v0.0.119.

* **Tests**
  * New unit, browser‑parity and conformance tests covering the JSON accessors; conformance suite now 80 programs.

* **Chores**
  * Package/library version bumped to v0.0.119.

### CODE REVIEW

File: HISTORY.md [hunk 1]
```diff
@@ -252,6 +252,7 @@ Stage 11 shifts focus from evaluation infrastructure to the standard library and
...
... between `wasmtime` and Node.js). String utilities: `string_chars` (canonical bridge from `String` to `Array<String>`; each 1-byte element is independently `$alloc`-ed to keep the GC mark phase honest about interior-pointer rooting), `string_lines` (Python `splitlines()` — `\n` / `\r\n` / `\r` terminators; trailing terminator does not add empty segment), `string_words` (Python `split()` — whitespace runs collapse, empty segments discarded), `string_reverse`, `string_trim_start`, `string_trim_end`, `string_pad_start`, `string_pad_end` (JavaScript `padStart`/`padEnd` — fill cycles left-to-right and is truncated to exactly the padding length; empty fill is a no-op). Character classification: `is_digit`, `is_alpha`, `is_alphanumeric`, `is_whitespace`, `is_upper`, `is_lower` (first-byte ASCII predicates; `false` for empty string). First-character case conversion: `char_to_upper`, `char_to_lower` (only the first byte is transformed; remaining bytes pass through — useful for title-casing). ASCII-only; Unicode-aware variants tracked separately. Eliminates the hand-rolled `byte == 48 or byte == 49 or ...` patterns and gives agents the string primitives that every non-trivial parser needs. |
```
[the remark this answers is not in the document]
Comment (aallan) [reply HISTORY.md]: @coderabbitai Fixed in commit 6da9f21 — rows swapped so v0.0.118 precedes v0.0.119 (oldest-to-newest).  An awk-swap mishap introduced a duplicate row momentarily; cleaned up with a second pass.

File: scripts/check_skill_examples.py [hunk 2]
```diff
@@ -63,96 +63,104 @@
...
+    1832: ("FRAGMENT", "Correct: if/else with braces (common mistakes)"),
 
     # String escapes — bare expression
-    1879: ("FRAGMENT", "Correct escape sequence examples, bare strings"),
+    1912: ("FRAGMENT", "Correct escape sequence examples, bare strings"),
 
     # Map/Set common mistakes — bare let bindings
-    1887: ("FRAGMENT", "Wrong: standalone map_new/set_new without type context"),
+    1920: ("FRAGMENT", "Wrong: standalone map_new/set_new without type context"),
     1893: ("FRAGMENT", "Correct: map_new/set_new with type context"),
```
[the remark this answers is not in the document]
Comment (aallan) [reply scripts/check_skill_examples.py]: @coderabbitai Fixed in commit 6da9f21 — ALLOWLIST now has no duplicate keys (verified via AST walk: `Total entries: 59, dups: 0`).  Subtle point worth noting: the four real code blocks are at four distinct line numbers (1845 / 1878 / 1893 / 1926) in the current SKILL.md, so the fix wasn't "delete the stale 1845/1893 entries" but rather "give each of the four blocks its own explicit entry" — previously two of the four were shadowed via last-write-wins because someone typo'd the same line number for two different blocks.  My earlier #366 addition at 1878/1926 was correct; folded those into the main Import / Map-Set sections and added comments explaining the history.

File: tests/test_browser.py [hunk 3]
```diff
@@ -1259,6 +1259,117 @@ def test_char_case(self, tmp_path: Path) -> None:
...
+        Some(@Int) -> IO.print("!"),
+        None -> IO.print("none")
+      }
+    }
+  }
+}
+'''
+        wasm_path, _ = _compile_vera(source, tmp_path)
+        node = _run_node(wasm_path)
+        assert node["stdout"] == "Alice,30,true,3.14,3,none,none"
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_browser.py]: @coderabbitai Fixed in commit 6da9f21.  `test_layer1_coercions` now exercises json_as_array (JArray construction), json_as_object (via json_parse to get a real Map), json_as_int on +inf and -inf.  `test_layer2_compound_accessors` now runs every Layer-2 accessor against a non-object Json (JArray) and asserts they all return None — the chain correctly preserves the underlying `json_get` None-on-non-JObject behaviour.

File: tests/test_codegen.py [hunk 4]
```diff
@@ -12291,6 +12291,360 @@ def test_string_words_content_via_join(self) -> None:
...
+  requires(true) ensures(true) effects(pure)
+{
+  match json_as_int(JNumber(infinity())) {
+    Some(@Int) -> 0,
+    None -> 1
+  }
+}
+"""
+        assert _run(src) == 1
+
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_codegen.py:12551]: @coderabbitai Fixed in commit 6da9f21, but went one step further than suggested: rather than just *testing* that finite overflow traps (which would have documented the limitation), I extended the `json_as_int` guard to cover **all four** `float_to_int` trap paths — NaN, +inf, -inf, and finite overflow `|f| >= 2^63`.  The guard is one extra `||` pair: `@Float64.0 >= 9223372036854775808.0 || @Float64.0 < -9223372036854775808.0`.  Accessor now returns None for every non-representable-as-Int input, matching the natural user expectation for a `*_as_int` coercion.  Two new tests pin the behaviour (|f| = 2^63 and |f| = 2^64); CHANGELOG, HISTORY, spec, and SKILL.md one-liner updated accordingly.

File: tests/test_codegen.py [hunk 5]
```diff
@@ -12291,6 +12291,360 @@ def test_string_words_content_via_join(self) -> None:
...
+{
+  -- JArray is a valid Json but not an object, so json_get_int returns None.
+  match json_get_int(JArray([JNumber(1.0)]), "any") {
+    Some(@Int) -> -1,
+    None -> 1
+  }
+}
+"""
+        assert _run(src) == 1
+
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_codegen.py:12764]: @coderabbitai Fixed in commit 6da9f21.  Three new tests in a dedicated "Import gating" section: two assert json_as_* / json_get_* accessors do NOT emit `(import "vera" "json_parse")` or `(import "vera" "json_stringify")` entries in the compiled WAT; one complementary test uses json_parse directly and asserts the import DOES appear.  The complementary test pins the direction of the invariant — if it ever fails without the others also failing, the check itself is wrong, not the implementation.

File: HISTORY.md [hunk 6]
```diff
@@ -253,6 +253,7 @@ Stage 11 shifts focus from evaluation infrastructure to the standard library and
...
+| v0.0.119 | 23 Apr | **JSON typed accessors** ([#366](https://github.com/aallan/vera/issues/366)) — eleven new accessor functions shipped as **pure-Vera prelude functions**, not WASM translators. Six Layer-1 type-coercion accessors (`json_as_string`, `json_as_number`, `json_as_bool`, `json_as_int`, `json_as_array`, `json_as_object`) return `Option<T>` from a `Json` value, `Some` when the constructor matches and `None` otherwise. Five Layer-2 compound field accessors (`json_get_string`, `json_get_number`, `json_get_bool`, `json_get_int`, `json_get_array`) collapse `json_get(obj, key)` and the matching `json_as_*` coercion into a single call — the pattern 90% of real API-consuming code wants. `json_as_int` guards every `float_to_int` trap path (NaN, ±infinity, and finite overflow when |f| >= 2^63), returning `None` for every non-representable-as-Int input. Eliminates the two-level pattern-match boilerplate every JSON API consumer would otherwise write. First non-WASM-translator stdlib addition of Stage 11 — proves the prelude-function path scales for ergonomic combinators that don't need inline WAT. |
```
[the remark this answers is not in the document]
Comment (aallan) [reply HISTORY.md]: @coderabbitai Fixed in commit d5192e3 — wrapped `|f| >= 2^63` in backticks, same treatment as the `||` fix we made last week for v0.0.118's row.  Markdown table renders correctly.

File: scripts/check_skill_examples.py [hunk 7]
```diff
@@ -63,96 +63,105 @@
...
+    1755: ("FRAGMENT", "Correct: expression with indices (not full fn)"),
+    1867: ("FRAGMENT", "Wrong: non-exhaustive match (missing None)"),
+    1845: ("FRAGMENT", "Wrong: non-exhaustive match (missing arm)"),
+
+    # Import syntax — intentionally unsupported.  The "import aliasing"
+    # and "import hiding" blocks are at four distinct line numbers
+    # (1878 / 1893), not duplicated at 1845 / 1893 as the ALLOWLIST
+    # originally claimed — that was a stale-dict-key issue that
+    # silently shadowed two of the four entries.  The four blocks
+    # are now allowlisted at their actual fence line numbers.
```
[the remark this answers is not in the document]
Comment (aallan) [reply scripts/check_skill_examples.py]: @coderabbitai Fixed in commit d5192e3 — rewrote the comment to list all four distinct line numbers (1845, 1878, 1883, 1893) rather than the misleading parenthetical that named only two.  Also pointed at the #511 fix for context so future maintainers see the duplicate-key history.

File: spec/09-standard-library.md [hunk 8]
```diff
@@ -1808,6 +1808,82 @@ is_whitespace("\t")  -- true
 
 **Verification:** all sixteen functions have Tier-1-verifiable signatures; their bodies fall to Tier 3 (runtime) verification because the SMT encoding does not yet model byte-level string operations. Their `requires(true)` / `ensures(true)` contracts are total, so Tier 3 reduces to runtime trap-freedom — every input is accepted.
 
+### 9.6.23 JSON Typed Accessors
+
+Vera provides eleven additional JSON accessor functions that eliminate the two-level pattern-match boilerplate (`match option ... { Some(@Json) -> match @Json.0 { JNumber(@Float64) -> ... } }`) that every JSON API consumer would otherwise write. Tracked in [#366](https://github.com/aallan/vera/issues/366).
+
+Unlike the rest of the chapter-9 built-ins, these are **pure-Vera prelude functions**, not WASM translators. The compiler injects them into every module that references `Json` values, alongside the existing `json_get` / `json_keys` / `json_type` combinators. No new host imports; no new WASM emit.
+
```
[the remark this answers is not in the document]
Comment (aallan) [reply spec/09-standard-library.md]: @coderabbitai Fixed in commit d5192e3.  Rewrote the claim to: "No new host imports and no dedicated WASM translator paths — but the accessors are still compiled as ordinary WASM functions via the standard AST-to-WAT pipeline when a module actually references them.  A module that never calls any of the eleven accessors pays zero compiled-WASM cost for them."  The terse "no new WASM emit" phrasing was technically wrong; this spells out the gating.

File: tests/test_browser.py [hunk 9]
```diff
@@ -1259,6 +1259,171 @@ def test_char_case(self, tmp_path: Path) -> None:
...
+        Some(@Map<String, Json>) -> IO.print("obj"),
+        None -> IO.print("?")
+      }
+  }
+}
+'''
+        wasm_path, _ = _compile_vera(source, tmp_path)
+        node = _run_node(wasm_path)
+        assert node["stdout"] == "hi,none,3.14,true,42,none,none,none,2,obj"
+
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_browser.py:1384]: @coderabbitai Fixed in commit d5192e3.  The browser test_layer1_coercions now exercises +2^63 (exclusive upper bound, returns None), -2^63 (inclusive lower bound, returns Some(INT64_MIN)), and the next representable Float64 below -2^63 (returns None), pinning the upper-exclusive/lower-inclusive asymmetry of WASM's i64 range under the browser runtime.  Ran into #475 bug 9 (`int_to_string(INT64_MIN)` negation overflow) when printing the -2^63 value — worked around with `bool_to_string(@Int.0 < 0)` plus a test-body comment; the bug itself is on the queue for the v0.0.120 cleanup PR.

File: tests/test_codegen.py [hunk 10]
```diff
@@ -12291,6 +12291,470 @@ def test_string_words_content_via_join(self) -> None:
...
 
 
+class TestJsonTypedAccessors:
+    """#366 — 11 new Json typed accessors shipped as pure-Vera prelude
+    functions (no new WASM translators).  Six Layer-1 type-coercion
+    accessors (Json → Option<T>) and five Layer-2 compound field
+    accessors (Json, String → Option<T> = json_get + json_as_T
+    composed).  ``json_as_int`` specifically guards against the
+    ``float_to_int`` trap on NaN/Infinity via ``float_is_nan`` /
+    ``float_is_infinite``.
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_codegen.py]: @coderabbitai Fixed in commit d5192e3.  Class docstring now enumerates all four guarded trap paths (NaN, +inf, -inf, finite overflow |f| >= 2^63) and notes the guard mechanism (float_is_nan + float_is_infinite + explicit bounds against ±9223372036854775808.0).  Full-protection semantics are now visible at first glance.

File: tests/test_codegen.py [hunk 11]
```diff
@@ -12291,6 +12291,470 @@ def test_string_words_content_via_join(self) -> None:
...
+  requires(true) ensures(true) effects(pure)
+{
+  -- -2^64 is well outside i64 range.
+  match json_as_int(JNumber(0.0 - 18446744073709551616.0)) {
+    Some(@Int) -> 0,
+    None -> 1
+  }
+}
+"""
+        assert _run(src) == 1
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_codegen.py:12550]: @coderabbitai Fixed in commit d5192e3 with a twist: there are TWO relevant -2^63 tests because the i64 range is asymmetric.  `test_json_as_int_boundary_minus_2_63_is_representable` pins that exactly -2^63 returns Some(INT64_MIN) (lower bound is inclusive).  `test_json_as_int_boundary_just_below_minus_2_63` pins that the next representable Float64 below -2^63 returns None.  First attempt used -2^63 - 1024 as the "just below" value — that's exactly halfway between representable doubles, which rounds to -2^63 via round-to-nearest-even, so the test wrongly expected None when Some was returned.  The true ulp-below value is -2^63 - 2048 (Float64 ulp at that magnitude is 2^11 = 2048).  Test comment documents the precision quirk.

File: HISTORY.md [hunk 12]
```diff
@@ -253,6 +253,7 @@ Stage 11 shifts focus from evaluation infrastructure to the standard library and
...
+| v0.0.119 | 23 Apr | **JSON typed accessors** ([#366](https://github.com/aallan/vera/issues/366)) — eleven new accessor functions shipped as **pure-Vera prelude functions**, not WASM translators. Six Layer-1 type-coercion accessors (`json_as_string`, `json_as_number`, `json_as_bool`, `json_as_int`, `json_as_array`, `json_as_object`) return `Option<T>` from a `Json` value, `Some` when the constructor matches and `None` otherwise. Five Layer-2 compound field accessors (`json_get_string`, `json_get_number`, `json_get_bool`, `json_get_int`, `json_get_array`) collapse `json_get(obj, key)` and the matching `json_as_*` coercion into a single call — the pattern 90% of real API-consuming code wants. `json_as_int` guards every `float_to_int` trap path (NaN, ±infinity, and finite overflow when `|f| >= 2^63`), returning `None` for every non-representable-as-Int input. Eliminates the two-level pattern-match boilerplate every JSON API consumer would otherwise write. First non-WASM-translator stdlib addition of Stage 11 — proves the prelude-function path scales for ergonomic combinators that don't need inline WAT. |
```
[the remark this answers is not in the document]
Comment (aallan) [reply HISTORY.md]: @coderabbitai Declining this one.  The established format across every Stage 11 release entry (v0.0.112 through v0.0.118, nine rows directly above this one) is a **detailed paragraph** describing what shipped, what was learned, and what's deferred.  Shortening v0.0.119 to a single line would break format consistency, not restore it — it would be the only terse entry in a table of detailed ones.  If we want to move to a one-line format, that's a separate pass across all Stage 11 rows in HISTORY.md, not a change targeted at v0.0.119 alone.  No change.

File: spec/09-standard-library.md [hunk 13]
```diff
@@ -1808,6 +1808,82 @@ is_whitespace("\t")  -- true
...
+  requires(true) ensures(true) effects(pure)
+
+public fn json_as_object(@Json -> @Option<Map<String, Json>>)
+  requires(true) ensures(true) effects(pure)
+```
+
+Each `json_as_*` returns `Some(value)` when the Json's constructor matches the requested type, `None` otherwise. The accessors are disjoint — at most one returns `Some` for any given Json value.
+
+`json_as_int` is the one asymmetric case: it applies `float_to_int` to the underlying `Float64`, truncating toward zero. `float_to_int` (aka WASM's `i64.trunc_f64_s`) traps on four inputs: NaN, ±infinity, and any finite float with `|f| >= 2^63` (outside the i64 range). `json_as_int` guards all four cases (`float_is_nan`, `float_is_infinite`, plus explicit range checks against ±9223372036854775808.0) and returns `None` for every non-representable-as-Int input.
+
```
[the remark this answers is not in the document]
Comment (aallan) [reply spec/09-standard-library.md]: @coderabbitai Fixed in commit 3cf5a13.  Rewrote §9.6.23's json_as_int paragraph to state the exact i64 range ([-2^63, 2^63) — closed-open), explain the asymmetry (two's-complement can represent INT64_MIN = -2^63 but not +2^63), and add an explicit note that `json_as_int(JNumber(-2^63))` correctly returns `Some(INT64_MIN)`.  The old "|f| >= 2^63" phrasing was technically wrong because it conflated f = -2^63 (representable) with f = +2^63 (not).

File: TESTING.md [hunk 14]
```diff
@@ -6,9 +6,9 @@ This is the single source of truth for Vera's testing infrastructure, coverage d
 
 | Metric | Value |
 |--------|-------|
-| **Tests** | 3,501 across 28 files (~34,000 lines of test code; 3,487 passed, 14 skipped) |
+| **Tests** | 3,534 across 28 files (~34,000 lines of test code; 3,520 passed, 14 skipped) |
 | **Compiler code coverage** | 96% of 15,149 statements (CI minimum: 80%) |
-| **Conformance programs** | 79 programs across 9 spec chapters, validating every language feature |
+| **Conformance programs** | 80 programs across 9 spec chapters, validating every language feature |
```
[the remark this answers is not in the document]
Comment (aallan) [reply TESTING.md:11]: @coderabbitai Fixed in commit 3cf5a13 — line 85 now reads "collection of 80 small, focused programs", matching the header at line 11.  `check_doc_counts.py` passes; no other "79" references in TESTING.md.

File: tests/test_codegen.py [hunk 15]
```diff
@@ -12291,6 +12291,520 @@ def test_string_words_content_via_join(self) -> None:
...
+{
+  match json_as_int(JNumber(0.0 - 9223372036854775808.0)) {
+    Some(@Int) -> 1,
+    None -> 0
+  }
+}
+"""
+        # Exactly representable — expect Some(INT64_MIN), so we see 1.
+        assert _run(src) == 1
+
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_codegen.py:12527]: @coderabbitai Fixed in commit 3cf5a13.  Upgraded `test_json_as_int_boundary_minus_2_63_is_representable` from a Some(_)-only check to two observable probes that together pin the value to INT64_MIN without hitting #475 bug 9 (int_to_string(INT64_MIN) negation overflow):

- Probe (a): `bool_to_string(@Int.0 < 0)` asserts "true" — catches accidental truncation or positive return.
- Probe (b): `int_to_string(@Int.0 + 1)` asserts "-9223372036854775807" (INT64_MIN + 1, which IS serialisable without the bug) — catches any value other than INT64_MIN exactly.

Assertion is now `_run_io(src) == "true,-9223372036854775807"`.  Same probe mirrored into the browser test's `test_layer1_coercions`.  Inline comment explains why the indirect probes are necessary pending the v0.0.120 #475 cleanup.

File: HISTORY.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (aallan) [reply HISTORY.md]: @coderabbitai Retracting my earlier dismissal — you were right, twice now.  I looked at Stage 11 entries only when I said "the established format is detailed paragraphs" and missed that Stage 1 through ~Stage 8 rows are one or two sentences.  The Stage 11 entries had drifted into essay-length prose, and I was defending the drift instead of the convention.

Fixed in commit 2a90405 — v0.0.119 row shortened from ~240 words to ~70, matching the early-stage format.  A broader pass over the Stage 11 rows to bring them all in line is worth doing but out of scope for this PR; I'll track it as a follow-up.

Thanks for the push-back.

File: tests/test_codegen.py [hunk 16]
```diff
@@ -12291,6 +12291,535 @@ def test_string_words_content_via_join(self) -> None:
...
 
 
+class TestJsonTypedAccessors:
+    """#366 — 11 new Json typed accessors shipped as pure-Vera prelude
+    functions (no new WASM translators).  Six Layer-1 type-coercion
+    accessors (Json → Option<T>) and five Layer-2 compound field
+    accessors (Json, String → Option<T> = json_get + json_as_T
+    composed).  ``json_as_int`` specifically guards against all four
+    ``float_to_int`` (i.e. ``i64.trunc_f64_s``) trap paths — NaN,
+    +infinity, -infinity, and finite overflow (|f| >= 2^63) — via
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_codegen.py]: @coderabbitai Fixed in commit 0db5ce2 — the class docstring now states the exact closed-open i64 range `[-2^63, 2^63)` and the precise trap condition (`f >= 2^63 or f < -2^63`), matching the boundary tests (`test_json_as_int_boundary_minus_2_63_is_representable` and `test_json_as_int_boundary_just_below_minus_2_63`).  Fourth iteration of this piece of wording across the cycle — spec text, CHANGELOG, docstring, and HISTORY all now converge on the same asymmetric-range description.

File: tests/test_codegen.py [hunk 17]
```diff
@@ -12291,6 +12291,535 @@ def test_string_words_content_via_join(self) -> None:
...
+public fn main(-> @Int)
+  requires(true) ensures(true) effects(pure)
+{
+  match json_as_int(JNumber(infinity())) {
+    Some(@Int) -> 0,
+    None -> 1
+  }
+}
+"""
+        assert _run(src) == 1
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_codegen.py]: @coderabbitai Fixed in commit 0db5ce2 — `test_json_as_int_infinity_returns_none` now exercises both +inf and -inf, with a docstring note that both are distinct trap inputs for `i64.trunc_f64_s` (separate from finite-overflow) and both are caught by the sign-agnostic `float_is_infinite` guard.  Browser parity test was already covering both signs.
